### PR TITLE
fix(chat): name the attachment a provider could never have fetched

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Symptom

Any attachment sent with a chat turn dies at the provider with a bare 400. The transport is fine — the upload returns 200, the stream POST returns 200 — the failure is the provider call itself, and nothing in the error says an attachment was involved. Reproduced twice through a real browser against a production-mode server with CSRF enabled, with two different attachments.

## Mechanism

An attachment reaches the provider as a **URL, not as bytes**. The OpenAI-compatible converter emits the URL verbatim and the provider's own servers dereference it:

```ts
// src/provider/runtime-loader.ts:213
content.push({ type: "image_url", image_url: { url: part.url } });
```

The chat upload handler mints a URL that never resolves from there. When the configured storage backend returns no external URL of its own, `POST` falls back to **this app's own origin**:

```ts
// src/chat/upload-handler.ts:426
const url = external ?? fallbackUrl(request.url, ref.id);
```

Driving the published `veryfront@0.1.1232` modules end to end — real upload handler, real converters:

```
1. upload response  : {"id":"blob_1","url":"https://app.example.com/api/chat/upload?id=blob_1",
                       "name":"screenshot.png","mediaType":"image/png","size":70}
3. OPENAI WIRE BODY : [{"role":"user","content":[
     {"type":"text","text":"what is in this image?"},
     {"type":"image_url","image_url":{"url":"https://app.example.com/api/chat/upload?id=blob_1"}},
     ...]}]
```

On a locally run production-mode server that origin is `http://localhost:PORT`, so the provider resolves it to itself and answers 400. The upload succeeded; the turn dies with nothing naming the file.

## Fix

Not fixable at that layer — the converter holds a URL, not bytes — so this takes the option the brief calls for: **fail with a message that says which attachment and why**, at the last point where both are still known (`getUserFileParts`, the provider-agnostic converter, so every provider gets the same message).

```
Attachment "screenshot.png" cannot be sent to the model: "localhost" is a loopback
address that resolves to the provider's own machine. The provider fetches attachments
from its own network, so the URL must be reachable from the public internet. ...
```

Only hosts unreachable **by construction** are rejected — loopback, link-local, RFC 1918 — plus schemes a provider cannot dereference (`blob:`, `file:`). `data:` URLs carry their own bytes and stay allowed. A public host that merely happens to be firewalled is not detectable, and guessing would break working setups, so it is left to the provider. Verified non-rejections include `172.32.0.1`, `192.169.1.1`, `11.0.0.1`, and `localhost.example.com`.

## Test

Red first, in the converter test:

```
names the attachment when its URL is one no provider can fetch ... FAILED (1ms)

error: AssertionError: Expected function to throw.
    at src/agent/runtime/text-generation-runtime-message-converter.test.ts:114:21
```

That is the defect exactly: the loopback upload URL is passed straight through to the provider today.

Green after, plus 14 unit steps in `attachment-reachability.test.ts` covering the boundary cases above.

Full `src/agent/` + `src/chat/`: `ok | 1211 passed (2078 steps) | 0 failed`. Lint, `deno check`, and the pre-push suite are green.

## Review follow-ups (d0ad10617)

Two findings from the Codex review, both real, both fixed.

**A server-local runtime is not "the provider".** The check answers "is this URL reachable from the public internet", and the first commit applied that answer to every runtime. A `server-local` runtime fetches from *this* machine, where the loopback URL the upload handler mints resolves fine. The built-in local runtime drops non-text parts anyway (`src/provider/local/model-runtime-adapter.ts:56`), so a local-model turn with an attachment used to answer text-only — the unconditional assertion turned that into a hard failure. Both call sites hold the `ModelRuntime` one line above the conversion, so they now pass `requireInternetReachableAttachments: !isLocalModelRuntime(languageModel)`. The option defaults to `true`, so remote keeps the check.

**IPv6 hosts skipped the check entirely.** `readIPv4Octets("[fc00::1]")` returned `undefined` and the host was waved through to the same opaque 400 this exists to replace. Now judged: `fc00::/7`, `fe80::/10`, `::1`, `::`, and IPv4-mapped addresses unwrapped and judged by the IPv4 rules (`URL` re-serializes `::ffff:127.0.0.1` to `::ffff:7f00:1`, so the hextets have to be decoded). Boundaries are pinned in both directions — `fbff::1`, `fe7f::1` and `fec0::1` are tested as *allowed* so the masks cannot quietly widen.

**Two more from the CodeRabbit pass (2a4b91d73).** A trailing DNS root dot bypassed the whole host check: `URL` keeps it on a name and normalizes it away on an IP literal, so `http://localhost./api/chat/upload?id=blob_1` arrived with hostname `localhost.` — not in the loopback set, not a suffix match, waved through. One trailing dot is now dropped before the host is judged. And the exemption above was only tested at the innermost converter, while the runtime enters three layers up; deleting the `...options` spread in the batch hand-off passed the entire suite, so the exemption could have stopped reaching the provider boundary in silence. All three entry points are now pinned in both directions.

Two suggestions were declined on the record, with evidence, in the threads: sanitizing the URL out of the error message (the error carries `attachmentUrl` by construction, so it removes nothing, and the reader is the user who uploaded it), and converting the error to `defineError` (four sibling classes in the same directory are plain `Error` subclasses — `resume-session.ts:7,15,23,31`).

## Related

The companion PR fixes why the log read `err=undefined:` — DNT rewriting `new.target` into the `import.meta` ponyfill. The two are independent; neither is based on the other.

Do not merge or queue — review only.

## Summary by CodeRabbit

- **New Features**
  - Added validation to ensure attachments sent to remote model providers use internet-reachable URLs.
  - Public and data URLs remain supported, while local, private, loopback, and unsupported URLs are identified.
  - Errors now clearly identify the affected attachment and unreachable address.

- **Bug Fixes**
  - Server-local runtimes can continue using attachments hosted on loopback or private network addresses.
  - Reachability checks apply consistently across single-message, batch, streaming, and request-based generation flows.

### CODE REVIEW

File: src/agent/runtime/text-generation-runtime-message-converter.ts [hunk 1]
```diff
@@ -210,6 +211,15 @@ function getUserFileParts(parts: Message["parts"]): TextGenerationRuntimeFilePar
...
     if (!mediaType || !url) return [];
 
+    // The provider dereferences this URL from its own network. A URL that can
+    // never resolve there comes back as a bare 400 that names nothing, so it
+    // fails here instead, naming the attachment.
+    assertProviderReachableAttachment({
+      url,
+      filename: getStringPartField(part, "filename"),
+      mediaType,
+    });
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/text-generation-runtime-message-converter.ts]: Fixed in d0ad10617 — you were right, and the regression is concrete.

`convertToTextGenerationRuntimeRequestMessages` is called from exactly two places (`src/agent/runtime/index.ts:1213` and `:1827`), and both have `languageModel` in hand, so the runtime *is* known one line above the conversion. Both now pass `requireInternetReachableAttachments: !isLocalModelRuntime(languageModel)`; the option defaults to `true` so a remote provider keeps the check.

Worth naming the exact regression, because it is not only the custom-runtime case you flagged: the built-in `server-local` runtime (`src/provider/local/model-runtime-adapter.ts:103`) drops non-text parts in `convertPrompt` (`:56` keeps only `part.type === "text"`), so before this PR a local-model turn with a loopback attachment answered text-only. With the unconditional assertion it died instead. That is a working setup broken by a check that had no business judging it — `assertProviderReachableAttachment` answers "reachable from the public internet", which is the wrong question for a runtime that fetches from this machine.

Covered by `text-generation-runtime-message-converter.test.ts` — "keeps a loopback attachment for a runtime that fetches from this machine" asserts the `localhost` upload URL survives conversion untouched when the flag is off.

File: src/agent/runtime/attachment-reachability.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  const host = hostname.toLowerCase();
+  if (LOOPBACK_HOSTNAMES.has(host)) {
+    return `"${hostname}" is a loopback address that resolves to the provider's own machine`;
+  }
+  if (LOCAL_SUFFIXES.some((suffix) => host.endsWith(suffix))) {
+    return `"${hostname}" is a local-network name that does not resolve on the public internet`;
+  }
+
+  const octets = readIPv4Octets(host);
+  if (!octets) return undefined;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/attachment-reachability.ts]: Fixed in d0ad10617. Confirmed the hole: `readIPv4Octets("[fc00::1]")` splits on `.`, gets one part, returns `undefined`, and the old early return waved the host through — so `http://[fc00::1]/x` reached the provider and produced exactly the opaque 400 this module exists to replace.

Now judged, in `describeUnreachableIPv6`: `fc00::/7` and `fe80::/10` as private-network, `::1` as loopback, `::` as the unspecified address. `URL` brackets, lowercases and compresses every IPv6 host, so `readIPv6Address` only has to strip the brackets — `http://[FE80::1]/x` arrives as `[fe80::1]`.

One case your comment did not list but which is the same bypass: IPv4-mapped addresses. `http://[::ffff:127.0.0.1]/a.png` is a loopback wearing an IPv6 coat, and `URL` re-serializes it to the hextet form `[::ffff:7f00:1]`, so the mapping is decoded back to octets and judged by the IPv4 rules. There is a test that pins that serialization so a future URL-parser change cannot silently reopen it.

Also removed `::1`/`[::1]` from `LOOPBACK_HOSTNAMES` — with bracketed hosts routed through the IPv6 path those entries were dead, and one authority per address family is easier to keep honest. `fbff::1` and `fec0::1` are pinned as *allowed* so the `fc00::/7` and `fe80::/10` masks cannot quietly widen.

File: src/agent/runtime/attachment-reachability.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        `The provider fetches attachments from its own network, so the URL must be reachable ` +
+        `from the public internet. Configure chat upload storage that serves a public URL, ` +
+        `or run behind a publicly reachable origin.`,
+    );
+    this.name = "UnreachableAttachmentError";
+    this.filename = options.filename;
+    this.attachmentUrl = options.attachmentUrl;
+    this.reason = options.reason;
+  }
+}
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/attachment-reachability.ts:63]: Three claims here; taking them separately, because one is right and two are not.

**Trailing DNS root dot — right, fixed in 2a4b91d73.** Reproduced before touching anything:

```
"localhost."       hostname= "localhost."       => ALLOWED
"app.localhost."   hostname= "app.localhost."   => ALLOWED
"nas.local."       hostname= "nas.local."       => ALLOWED
"127.0.0.1."       hostname= "127.0.0.1"        => ... is a loopback address ...
```

`URL` keeps the root dot on a DNS name and normalizes it away on an IP literal, which is exactly why this slipped past both the set lookup and the suffix match while the IPv4 path was unaffected. One trailing dot is now dropped before the host is judged, with `localhost.`, `app.localhost.`, `nas.local.` and `svc.internal.` pinned as rejected. Reverting the strip fails the new test, so it is red-first.

**Sanitizing the URL out of the message — declining.** It would not reduce exposure. `UnreachableAttachmentError` carries the full URL as a field by construction (`attachment-reachability.ts:38`, `attachmentUrl: options.url`) — that is the point of the class, so anything reading the error already has it. Every other reason interpolates the host too (`:80`, `:98`), so removing it from one branch buys nothing and makes the least self-explanatory failure the least debuggable. And the trust boundary does not support the premise: this is the user's own upload URL, minted by our upload handler for their attachment, surfaced back to that same user in the turn they sent it. Happy to revisit if you can name a reader of the message that is not already a reader of `attachmentUrl`.

**`defineError` — declining, not this directory's convention.** Four sibling classes in the same directory are plain `Error` subclasses: `src/agent/runtime/resume-session.ts:7,15,23,31` (`RunCancelledError`, `RunAlreadyExistsError`, `RunNotActiveError`, `WaitNotPendingError`). Same across `src/agent/` — `src/agent/input/human-input.ts:202,212`, `src/agent/hosted/context-budget-manager.ts:101`, `src/agent/hosted/child-status.ts:72`. `defineError` is used by a handful of modules (`src/extensions/`, `src/integrations/remote-tools.ts`, `src/transforms/mdx/...`), and registering a new error also moves the error-registry count assertions. Converting one class here would make this file the odd one out and put a registry change in a chat-attachment PR.

File: src/agent/runtime/text-generation-runtime-message-converter.test.ts [hunk 4]
```diff
@@ -86,6 +86,68 @@ describe("text-generation-runtime-message-converter", () => {
...
+
+      const content = (result as TextGenerationRuntimeUserMessage).content;
+      if (!Array.isArray(content)) {
+        throw new Error("Expected user content to preserve native file parts");
+      }
+      const fileUrls = content.flatMap((part) =>
+        part.type === "file" || part.type === "image" ? [part.url] : []
+      );
+      assertEquals(fileUrls, ["http://localhost:3000/api/chat/upload?id=blob_1"]);
+    });
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/agent/runtime/text-generation-runtime-message-converter.test.ts:150]: Right, and it was not hypothetical — fixed in 2a4b91d73.

The runtime calls `convertToTextGenerationRuntimeRequestMessages`, which delegates to `convertToTextGenerationRuntimeMessages`, which delegates to `convertToTextGenerationRuntimeMessage`. The exemption only reaches the assertion through the `...options` spread in the middle hand-off (`text-generation-runtime-message-converter.ts:502`). I deleted that spread and re-ran the suite: everything still passed. So the option could have stopped reaching the provider boundary with the tests green, and the server-local regression this PR just fixed would have come straight back.

Both outer entry points are now pinned in both directions — exemption honoured with `requireInternetReachableAttachments: false`, attachment still rejected by default:

```
convertToTextGenerationRuntimeMessages forwards the server-local exemption ... ok
convertToTextGenerationRuntimeMessages still rejects the attachment by default ... ok
convertToTextGenerationRuntimeRequestMessages forwards the server-local exemption ... ok
convertToTextGenerationRuntimeRequestMessages still rejects the attachment by default ... ok
```

Confirmed red-first: with the `...options` spread removed, both `forwards the server-local exemption` cases fail.
